sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py

```diff
@@ -0,0 +1,179 @@
+      # This timestamp is used as the monotonic clock to order events in the
+      # replay.
+      self._monotonic_clock = timestamp.Timestamp.of(0)
+
+      # The maximum timestamp read.
+      self._target_timestamp = timestamp.Timestamp.of(0)
+
+      # The PCollection cache readers.
+      self._readers = {}
+
+      # The file headers that are metadata for that particular PCollection.
+      self._headers = {}
+
+      # The header allows for metadata about an entire stream, so that the data
+      # isn't copied per record.
+      readers = [r.read() for r in readers]
+      for r in readers:
+        header = InteractiveStreamHeader()
+        header.ParseFromString(next(r))
+
+        # Main PCollections in Beam have a tag as None. Deserializing a Proto
+        # with an empty tag becomes an empty string. Here we normalize to what
+        # Beam expects.
+        self._headers[header.tag if header.tag else None] = header
+        self._readers[header.tag if header.tag else None] = r
+
+      # The watermarks per tag. Useful for introspection in the stream.
+      self._watermarks = {tag: timestamp.MIN_TIMESTAMP for tag in self._headers}
+
+      # The most recently read timestamp per tag.
+      self._stream_times = {tag: timestamp.MIN_TIMESTAMP
+                            for tag in self._headers}
+
+    def _read_next(self):
+      """Reads the next iteration of elements from each stream.
+      """
+      records = []
+      for tag, r in self._readers.items():
+        # The target_timestamp is the maximum timestamp that was read from the
+        # stream. Some readers may have elements that are less than this. Thus,
+        # we skip all readers that already have elements that are at this
+        # timestamp so that we don't read everything into memory.
+        if self._stream_times[tag] >= self._target_timestamp:
+          continue
+        try:
+          record = InteractiveStreamRecord()
+          record.ParseFromString(next(r))
+          records.append((tag, record))
+          self._stream_times[tag] = Timestamp.from_proto(
+              record.processing_time)
+        except StopIteration:
+          pass
+      return records
+
+    def read(self):
+      """Reads records from PCollection readers.
+      """
+      # We use a generator here because the underlying readers may have to much
+      # data to read into memory.
+
+      events = []
+      while True:
+        # Read the next set of events. The read events will most likely be
+        # out of order if there are multiple readers. Here we sort them into
+        # a more manageable state.
+        events = events + self._read_next()
+        events = sorted(events,
+                        key=lambda x: Timestamp.from_proto(
+                            x[1].processing_time),
+                        reverse=True)
+        if not events:
+          break
+
+        # Retrieves the minimum timestamp from the read events.
+        min_timestamp = (
```

Comment:
Why is this a callable?

Reply:
It's callable because it retrieves the the minimum timestamp of the events. Made it clearer by adding events as an argument and into its own method.